cmd/vela-email/plugin.go

```diff
@@ -0,0 +1,269 @@
+			envMap[splitV[0]] = strings.Join(splitV[1:], "=")
+		}
+	}
+
+	envMap["BuildCreated"] = p.BuildEnv.BuildCreated
+	envMap["BuildEnqueued"] = p.BuildEnv.BuildEnqueued
+	envMap["BuildFinished"] = p.BuildEnv.BuildFinished
+	envMap["BuildStarted"] = p.BuildEnv.BuildStarted
+
+	return envMap
+}
+
+// Parses subject and body of email to inject environment
+// variables. Uses provided authentication type and send type
+// to send the email.
+func (p *Plugin) Exec() error {
+	logrus.Trace("entered plugin.Execute")
+	defer logrus.Trace("exited plugin.Execute")
+
+	logrus.Debug("Parsing Subject...")
+	subject, err := p.injectEnv(p.Email.Subject)
+	if err != nil {
+		return err
+	}
+	p.Email.Subject = subject
+
+	if len(p.Email.HTML) > 0 {
+		logrus.Debug("Parsing HTML...")
+		body, err := p.injectEnv(string(p.Email.HTML))
+		if err != nil {
+			return err
+		}
+		logrus.Debug("Parsing CSS...")
+		body, err = inliner.Inline(body)
+		if err != nil {
+			return err
+		}
+		p.Email.HTML = []byte(body)
+	} else {
+		logrus.Debug("Parsing Text...")
+		body, err := p.injectEnv(string(p.Email.Text))
+		if err != nil {
+			return err
+		}
+		p.Email.Text = []byte(body)
+	}
+
+	var auth smtp.Auth
+	switch strings.ToLower(p.Auth) {
+	case "plainauth":
+		logrus.Info("Using login authentication from smtp/PlainAuth...")
+		auth = smtp.PlainAuth("", p.SMTPHost.Username, p.SMTPHost.Password, p.SMTPHost.Host)
+	case "loginauth":
+		logrus.Info("Using login authentication from loginauth/LoginAuth...")
+		auth = LoginAuth(p.SMTPHost.Username, p.SMTPHost.Password)
+	default:
+		logrus.Info("Using no login authentication...")
+		auth = nil
+	}
+
+	host := p.SMTPHost.Host + ":" + p.SMTPHost.Port
+	switch strings.ToLower(p.SendType) {
+	case "starttls":
+		logrus.Info("Sending email with StartTLS...")
+		if err := p.Email.SendWithStartTLS(host, auth, p.TLSConfig); err != nil {
+			return fmt.Errorf("error sending with StartTLS: %w", err)
+		}
+	case "tls":
+		logrus.Info("Sending email with TLS...")
+		if err := p.Email.SendWithTLS(host, auth, p.TLSConfig); err != nil {
+			return fmt.Errorf("error sending with TLS: %w", err)
+		}
+	case "plain":
+		fallthrough
+	default:
+		logrus.Info("Sending email with Plain...")
+		if err := p.Email.Send(host, auth); err != nil {
+			return fmt.Errorf("error sending with Plain: %w", err)
+		}
+	}
+
+	logrus.Info("Plugin finished")
+	return nil
+}
+
+// Injects environment variables into email template.
+func (p *Plugin) injectEnv(str string) (string, error) {
+	logrus.Trace("entered plugin.InjectEnv")
+	defer logrus.Trace("exited plugin.InjectEnv")
+
+	// Inject to subject
+	buffer := new(bytes.Buffer)
+
+	// parse string to template
+	t := template.Must(template.New("input").Parse(str))
+
+	err := t.Execute(buffer, p.Env)
+
+	return buffer.String(), err
+}
+
+// splits a string of emails and returns them as a slice.
+func stringToSlice(s []string) []string {
+	var slice []string
+
+	if len(s) == 0 {
```

Comment:
I think this length check4 code isn't needed. If you range over a length 0 slice it shouldn't have anything in it. So you should be able to remove this code without impacting any of the function or logic.